src/Core/Framework/DataAbstractionLayer/Search/RequestCriteriaBuilder.php

```diff
@@ -510,4 +513,59 @@ private function buildFieldName(EntityDefinition $definition, string $fieldName)
 
         return $fieldName;
     }
+
+    /**
+     * @return array<string, mixed>
+     */
+    private function decodeCriteriaParameter(string $encodedCriteria): array
+    {
+        try {
+            // Hard limit to avoid overloading
+            if (\strlen($encodedCriteria) > self::MAX_GET_CRITERIA_LENGTH) {
+                throw DataAbstractionLayerException::invalidCriteriaParameter('The _criteria parameter is too long');
+            }
+
+            // Decode base64url
+            $gzippedData = $this->base64urlDecode($encodedCriteria);
+            if ($gzippedData === false) {
+                throw DataAbstractionLayerException::invalidCriteriaParameter('Unable to decode base64 data');
+            }
+
+            // Decompress gzipped data
+            // Limit the decompressed size for additional safety from malicious input.
+            // Function throws a warning on failure, suppressing it as result is validated afterwards.
+            $jsonData = @gzdecode($gzippedData, self::MAX_GET_CRITERIA_UNCOMPRESSED_LENGTH);
+
+            if ($jsonData === false) {
+                throw DataAbstractionLayerException::invalidCriteriaParameter('Unable to decompress gzipped data');
+            }
+
+            // Decode JSON
+            $criteriaData = json_decode($jsonData, true, 512, \JSON_THROW_ON_ERROR);
+
+            if (!\is_array($criteriaData)) {
+                throw DataAbstractionLayerException::invalidCriteriaParameter('Criteria data must be an array');
+            }
+
+            return $criteriaData;
+        } catch (\JsonException $e) {
+            throw DataAbstractionLayerException::invalidCriteriaParameter('Invalid JSON data: ' . $e->getMessage());
+        } catch (DataAbstractionLayerException $e) {
+            throw $e;
+        } catch (\Throwable $e) {
+            throw DataAbstractionLayerException::invalidCriteriaParameter('Unable to decode or decompress criteria data: ' . $e->getMessage());
+        }
+    }
+
+    // The standard base64 alphabet contains + and / which are not URL safe.
+    // Base64url encoding replaces + with - and / with _
+    // (see RFC 4648, Section 5 https://datatracker.ietf.org/doc/html/rfc4648#section-5).
+    // Padding restoration is unnecessary, as base64_decode handles it correctly.
+    private function base64urlDecode(string $data): string|false
```

Comment:
crazy that this is not yet a build in php method, maybe something for an RFC there

But anyway seems like even symfony etc is doing this in a couple of places, but does not offer it in a reusable way, maybe we should directly put that in another util so it is reusable OOTB